engines/docker/image_cache.go

```diff
@@ -0,0 +1,103 @@
+package dockerengine
+
+import (
+	"context"
+	"strings"
+	"time"
+
+	docker "github.com/fsouza/go-dockerclient"
+	"github.com/taskcluster/taskcluster-worker/runtime/atomics"
+	"github.com/taskcluster/taskcluster-worker/runtime/caching"
+	"github.com/taskcluster/taskcluster-worker/runtime/gc"
+)
+
+const dockerPullImageInactivityTimeout = 30 * time.Second
+
+type imageResource struct {
+	Repository string
+	Tag        string
+	Size       uint64
+	dispose    atomics.Once
+}
+
+type cacheContext struct {
+	context.Context
+}
+
+func newCacheContext(ctx context.Context) *cacheContext {
+	if ctx != nil {
+		return &cacheContext{
+			Context: ctx,
+		}
+	}
+	return nil
+}
+
+func (ctx *cacheContext) Progress(description string, percent float64) {
+	debug("progress: %s %f", description, percent)
+}
+
+func (im *imageResource) MemorySize() (uint64, error) {
+	return im.Size, nil
+}
+
+func (im *imageResource) DiskSize() (uint64, error) {
+	return 0, gc.ErrDisposableSizeNotSupported
+}
+
+func (im *imageResource) Dispose() error {
+	im.dispose.Do(func() {
+		client, err := docker.NewClientFromEnv()
+		if err != nil {
+			debug("%v", err)
+			return
+		}
+		err = client.RemoveImage(buildImageName(im.Repository, im.Tag))
+		if err != nil {
+			debug("%v", err)
+		}
+	})
+	return nil
+}
+
+func imageConstructor(ctx caching.Context, opts interface{}) (caching.Resource, error) {
+	options := opts.(imageType)
+	client, err := docker.NewClientFromEnv()
```

Comment:
Ah, I see, you think you can't pass in properties like `*engine` through `imageType`, you can, as long as it has an annotation that makes `json.Marshal` ignore it...

Reply:
Example:
```go
// imageOptions are passed to imageConstructor through *caching.Cache to provide options for creating an image
type imageOptions struct {
  Image  imageType `json:"image"` // options from task.payload
  Engine *engine   `json:"-"`     // engine reference w. monitor and docker client that's valid for the life-cycle of the image 
}
```

So long as we use `json:"-"` the property will be ignored by `json.Marshal` so when `caching.Cache` computes a hash value for the image (in-order to avoid duplicate images), it won't hash the value of the `*engine` pointer (which would be a true mess).

Reply:
This trick is useful for passing in references to things like docker client, and monitor that we might want to use later.. But that doesn't determine the identity of the image being downloaded..